auth/env_keychain_test.go

```diff
@@ -17,125 +17,112 @@ import (
+				h.AssertNil(t, os.Unsetenv("CNB_REGISTRY_AUTH"))
 			})
-		})
-
-		when("#EnvKeychain", func() {
-			var envKeyChain authn.Keychain
 
-			when("#Resolve", func() {
-				it.Before(func() {
-					envKeyChain = &auth.EnvKeychain{EnvVar: "CNB_REGISTRY_AUTH"}
-				})
-
-				it.After(func() {
-					err := os.Unsetenv("CNB_REGISTRY_AUTH")
-					h.AssertNil(t, err)
-				})
-
-				when("valid auth env variable is set", func() {
-					it.Before(func() {
-						err := os.Setenv(
-							"CNB_REGISTRY_AUTH",
-							`{"basic-registry.com": "Basic asdf=", "bearer-registry.com": "Bearer qwerty="}`,
-						)
+			when("valid auth env variable is set", func() {
+				when("valid variable", func() {
+					it("loads the basic auth from memory", func() {
+						resolvedKeychain, err := auth.ResolveKeychain("CNB_REGISTRY_AUTH", []string{})
 						h.AssertNil(t, err)
-					})
+						h.AssertNil(t, os.Unsetenv("CNB_REGISTRY_AUTH")) // unset variable to prove that our test is using a pre-resolved keychain
 
-					it("loads the basic auth from the environment", func() {
 						registry, err := name.NewRegistry("basic-registry.com", name.WeakValidation)
 						h.AssertNil(t, err)
 
-						authenticator, err := envKeyChain.Resolve(registry)
+						authenticator, err := resolvedKeychain.Resolve(registry)
 						h.AssertNil(t, err)
 
 						header, err := authenticator.Authorization()
 						h.AssertNil(t, err)
 
-						h.AssertEq(t, header, &authn.AuthConfig{Auth: "asdf="})
+						h.AssertEq(t, header, &authn.AuthConfig{Auth: "some-basic-auth="})
 					})
 
-					it("loads the bearer auth from the environment", func() {
+					it("loads the bearer auth from memory", func() {
+						resolvedKeychain, err := auth.ResolveKeychain("CNB_REGISTRY_AUTH", []string{})
+						h.AssertNil(t, err)
+						h.AssertNil(t, os.Unsetenv("CNB_REGISTRY_AUTH")) // unset variable to prove that our test is using a pre-resolved keychain
+
 						registry, err := name.NewRegistry("bearer-registry.com", name.WeakValidation)
 						h.AssertNil(t, err)
 
-						authenticator, err := envKeyChain.Resolve(registry)
+						authenticator, err := resolvedKeychain.Resolve(registry)
 						h.AssertNil(t, err)
 
 						header, err := authenticator.Authorization()
 						h.AssertNil(t, err)
 
-						h.AssertEq(t, header, &authn.AuthConfig{RegistryToken: "qwerty="})
+						h.AssertEq(t, header, &authn.AuthConfig{RegistryToken: "some-bearer-auth="})
 					})
 
-					it("returns an Anonymous authenticator when the environment does not have a auth header", func() {
-						envKeyChain = auth.NewKeychain("CNB_REGISTRY_AUTH")
-						registry, err := name.NewRegistry("no-env-auth-registry.com", name.WeakValidation)
-						h.AssertNil(t, err)
+					when("the environment does not have an auth header", func() {
+						it("uses the fallback keychain", func() {
+							useFallbackKeychain(t)
+						})
+
+						when("the fallback keychain does not have an auth header", func() {
```

Comment:
IMO our use of `fallback keychain` here is a bit confusing, that is just a name that we thought of because we had too many variables in our function but it doesn't have any real semantic meaning... I think "default keychain" would be clearer.

Reply:
I'm fine with the `fallback keychain` term since we're falling to this keychain when the environment keychain doesn't work.
The problem that I have with the `default keychain` term is that there is a `DefaultKeychain` struct in `authn` and we're not using it here so I think it might be confusing.
What do you think?

Reply:
After talking about this more we are going with default keychain for reasons.